contracts/Eip712Forwarder.sol

```diff
@@ -0,0 +1,150 @@
+        bytes32 domainSeparator, bytes32 requestTypeHash, bytes memory suffixData, bytes memory sig) public {
+        _verifyNonce(req);
+        _verifySig(req, domainSeparator, requestTypeHash, suffixData, sig);
+        _updateNonce(req);
+
+        // solhint-disable-next-line avoid-low-level-calls
+        (bool success,) = req.target.call{gas : req.gasLimit}(abi.encodePacked(req.encodedFunction, req.senderAddress));
+        if (!success) {
+            // solhint-disable-next-line no-inline-assembly
+            assembly {// This assembly ensure the revert contains the exact string data
+                let returnDataSize := returndatasize()
+                returndatacopy(0, 0, returnDataSize)
+                revert(0, returnDataSize)
+            }
+        }
+    }
+
+
+    function _verifyNonce(ForwardRequest memory req) internal view {
+        require(nonces[req.senderAddress] == req.senderNonce, "nonce mismatch");
+    }
+
+    function _updateNonce(ForwardRequest memory req) internal {
+        nonces[req.senderAddress]++;
+    }
+
+    //register a requestTypeHash
+    // the given requestType must have GENERIC_PARAMS as the beginning of the type string,
+    // otherwise it can't be registered.
+
+    /**
+     * Register a new Request typehash.
+     * @param typeName - the name of the request type.
+     * @param extraParams - params to add to the request type, after initial "_ForwardRequest request" param
+     * @param subTypes - subtypes used by the extraParamsg
+     * @param subTypes2 - more subtypes, if sorted after _ForwardRequest (e.g. if type starts with lowercase)
+     */
+    function registerRequestType(string calldata typeName, string calldata extraParams, string calldata subTypes, string calldata subTypes2) external {
+
+        require(bytes(typeName).length > 0, "invalid typeName");
+        bytes memory types = abi.encodePacked(subTypes, GENERIC_TYPE, subTypes2);
+        string memory comma = bytes(extraParams).length > 0 ? "," : "";
+
+        bytes memory requestType = abi.encodePacked(
+            typeName, "(",
+            GENERIC_PARAMS, comma, extraParams, ")",
+            types
+        );
+        bytes32 requestTypehash = keccak256(bytes(requestType));
+        uint len = bytes(subTypes).length;
+        //sanity: avoid redefining our type, e.g.: subType="_ForwardRequest(whatever)_z"
+        require(len == 0 || bytes(subTypes)[len - 1] == ")", "invalid subType");
+        //sanity: parameters should not end parameters block
+        for (uint i = 0; i < bytes(extraParams).length; i++) {
+            require(bytes(extraParams)[i] != ")", "invalid extraParams");
+        }
+        require(!typeHashes[requestTypehash], "typehash already registered");
+        typeHashes[requestTypehash] = true;
+        emit RequestTypeRegistered(requestTypehash, string(requestType));
+    }
+
+    function isRegisteredTypehash(bytes32 typehash) public view returns (bool) {
```

Comment:
How is this method expected to be used?

Reply:
Should it be part of '_verifySig' on-chain?

Reply:
its there for completeness. you can re-register at will, and it will ignore previous run (since they both do the same: save boolean in a hash) I added it because I don't like to rely on public members - these are exposing internal structs to tests at best, and can't be considered an API.